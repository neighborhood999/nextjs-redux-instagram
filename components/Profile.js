import React from 'react';
import PropTypes from 'prop-types';
import RelativeTime from 'react-relative-time';
import {
  compose,
  mapProps,
  lifecycle,
  onlyUpdateForPropTypes,
  setPropTypes
} from 'recompose';
import { Card, Divider, Grid, Icon, Image, Loader } from 'semantic-ui-react';
import { chunk } from '../utils';

const Profile = ({ ProfileSection, PhotosSection }) =>
  <div>
    {ProfileSection}
    <Divider />
    {PhotosSection}
  </div>;

const withProfileSection = ({ user }) => {
  const {
    username,
    profile_picture,
    bio,
    full_name,
    website
  } = user.userDetails;
  let { counts } = user.userDetails;

  if (typeof counts === 'undefined') {
    counts = {};
  }

  return user.isDone
    ? <Grid style={{ marginTop: '50px' }} stackable>
        <Grid.Row>
          <Grid.Column width={4} />
          <Grid.Column width={8}>
            <Image circular floated="left" src={profile_picture} />
            <h3>{username} <small>{full_name}</small></h3>
            <h4>
              {counts.media} 則貼文{' | '}
              {counts.followed_by} 位粉絲{' | '}
              追蹤 {counts.follows} 人
            </h4>
            <small>{bio}</small>
          </Grid.Column>
          <Grid.Column width={4} />
        </Grid.Row>
      </Grid>
    : <Loader
        style={{ marginTop: '50px', paddingTop: '100px' }}
        active
        inline="centered"
      />;
};

const withPhotosSection = ({ user }) => {
  const photos = user.photos.map(photo =>
    <Grid.Column key={photo.id}>
      <Card>
        <Card.Content>
          <Image avatar src={photo.user.profile_picture} />{' '}
          {photo.user.username}
          <Card.Meta className="right floated">
            <RelativeTime value={photo.created_time * 1000} />
          </Card.Meta>
        </Card.Content>
        <Image src={photo.images.standard_resolution.url} />
        <Card.Content>
          <Icon name="comment" /> {photo.comments.count} 則留言
          <span className="right floated">
            <Icon name="heart" /> {photo.likes.count} 個讚
          </span>
        </Card.Content>
      </Card>
    </Grid.Column>
  );

  const GridRowPhotos = chunk(photos, 4).map((group, id) =>
    <Grid.Row key={id} columns={4}>
      {group}
    </Grid.Row>
  );

  return (
    <Grid style={{ marginTop: '50px' }} stackable>
      {user.isDonePhotos ? GridRowPhotos : <Loader active inline="centered" />}
    </Grid>
  );
};

const propTypes = {
  user: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired
};

export default compose(
  lifecycle({
    componentDidMount() {
      if (!this.props.user.userDetails.hasOwnProperty('id')) {
        const { requestUserAndPhotos } = this.props.actions;
        requestUserAndPhotos();
      }
    }
  }),
  onlyUpdateForPropTypes,
  setPropTypes(propTypes),
  mapProps(({ user }) => ({
    ProfileSection: withProfileSection({ user }),
    PhotosSection: withPhotosSection({ user })
  }))
)(Profile);
