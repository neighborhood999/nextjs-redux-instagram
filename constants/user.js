// @flow

export const REQUEST_USER_AND_PHOTOS = 'REQUEST_USER_AND_PHOTOS';
export const RECEIVE_USER = 'RECEIVE_USER';
export const RECEIVE_SELF_PHOTOS = 'RECEIVE_SELF_PHOTOS';
