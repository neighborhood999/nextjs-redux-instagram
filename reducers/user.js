// @flow

import type { UserState, Action } from '../types';
import {
  REQUEST_USER_AND_PHOTOS,
  RECEIVE_USER,
  RECEIVE_SELF_PHOTOS
} from '../constants/user';

const auth = (
  state: UserState = {
    isFetching: false,
    isFetchingPhotos: false,
    isDone: false,
    isDonePhotos: false,
    userDetails: {},
    photos: []
  },
  action: Action
) => {
  switch (action.type) {
    case REQUEST_USER_AND_PHOTOS:
      return {
        ...state,
        isFetching: true,
        isFetchingPhotos: true
      };
    case RECEIVE_USER:
      return {
        ...state,
        isFetching: false,
        isDone: true,
        userDetails: action.info
      };
    case RECEIVE_SELF_PHOTOS:
      return {
        ...state,
        isFetchingPhotos: false,
        isDonePhotos: true,
        photos: action.photos
      };
    default:
      return state;
  }
};

export default auth;
