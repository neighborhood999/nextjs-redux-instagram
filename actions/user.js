// @flow

import type { UserAction } from '../types';
import * as UserConstants from '../constants/user';

export const requestUserAndPhotos = (): UserAction => ({
  type: UserConstants.REQUEST_USER_AND_PHOTOS
});

export const receiveUser = (info: {}): UserAction => ({
  type: UserConstants.RECEIVE_USER,
  info
});

export const receiveSelfPhotos = (photos: Array<{}>): UserAction => ({
  type: UserConstants.RECEIVE_SELF_PHOTOS,
  photos
});
