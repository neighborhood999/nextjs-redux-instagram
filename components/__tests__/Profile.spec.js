import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import objectAssignDeep from 'object-assign-deep';
import Profile from '../Profile';
import { mockUser, mockPhotos } from '../../actions/__tests__/user.spec';

const setup = nextProps => {
  const props = objectAssignDeep(
    {
      user: {
        isFetching: false,
        isFetchingPhotos: false,
        isDone: false,
        isDonePhotos: false,
        userDetails: {},
        photos: []
      },
      actions: {
        requestUserAndPhotos: jest.fn(),
        receiveUser: jest.fn(),
        receiveSelfPhotos: jest.fn()
      }
    },
    nextProps
  );

  const component = mount(<Profile {...props} />);

  return {
    component,
    props
  };
};

describe('Profile Component', () => {
  test('Profile should be wrapped up in in lifecycle', () => {
    expect(Profile.displayName).toBe(
      'lifecycle(onlyUpdateForPropTypes(mapProps(Profile)))'
    );
  });

  test('ProfileSection and PhotosSection should be render Loader when fetching api', () => {
    const { component, props } = setup();

    expect(component.find('Loader')).toHaveLength(2);
    expect(component.find('Divider')).toHaveLength(1);
  });

  test('should called requestUserAndPhotos when componentDidMount', () => {
    const { props } = setup();

    expect(props.actions.requestUserAndPhotos).toBeCalled();
  });

  test('should render ProfileSection', () => {
    const { component, props } = setup({
      user: { isDone: true, userDetails: mockUser.data }
    });

    expect(component.find('h3').text()).toBe('P.J neighborhood999');
    expect(component.find('h4').text()).toBe('999 則貼文 | 999 位粉絲 | 追蹤 999 人');
    expect(component.find('small').at(1).text()).toBe(
      'A guy who write some code.'
    );

    const tree = renderer.create(<Profile {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render PhotosSection', () => {
    const { component, props } = setup({
      user: {
        isDonePhotos: true,
        userDetails: mockUser.data,
        photos: mockPhotos.data
      }
    });

    expect(component.find('Card')).toHaveLength(1);
    expect(component.find('CardContent').at(1).find('span').text()).toBe(
      ' 40 個讚'
    );

    const tree = renderer.create(<Profile {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
