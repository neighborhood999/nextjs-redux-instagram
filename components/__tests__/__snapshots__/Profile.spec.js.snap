// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component should render PhotosSection 1`] = `
<div>
  <div
    className="ui active centered inline loader"
    style={
      Object {
        "marginTop": "50px",
        "paddingTop": "100px",
      }
    }
  />
  <div
    className="ui divider"
  />
  <div
    className="ui stackable grid"
    style={
      Object {
        "marginTop": "50px",
      }
    }
  >
    <div
      className="four column row"
    >
      <div
        className="column"
      >
        <div
          className="ui card"
          href={undefined}
          onClick={[Function]}
        >
          <div
            className="content"
          >
            <img
              alt={undefined}
              className="ui avatar image"
              height={undefined}
              src="https://scontent.cdninstagram.com/t51.2885-19/s150x150/test1_a.jpg"
              width={undefined}
            />
             
            bivinity
            <div
              className="meta right floated"
            >
              <time
                dateTime="2017-06-07T11:43:10.000Z"
                title="2017-06-07T11:43:10.000Z"
              >
                6 months ago
              </time>
            </div>
          </div>
          <img
            alt={undefined}
            className="ui image"
            height={undefined}
            src="https://scontent.cdninstagram.com/t51.2885-15/s640x640/sh0.08/e35/test1_n.jpg"
            width={undefined}
          />
          <div
            className="content"
          >
            <i
              aria-hidden="true"
              className="comment icon"
            />
             
            4
             則留言
            <span
              className="right floated"
            >
              <i
                aria-hidden="true"
                className="heart icon"
              />
               
              40
               個讚
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Profile Component should render ProfileSection 1`] = `
<div>
  <div
    className="ui stackable grid"
    style={
      Object {
        "marginTop": "50px",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="four wide column"
      />
      <div
        className="eight wide column"
      >
        <img
          alt={undefined}
          className="ui left floated image"
          height={undefined}
          shape="circular"
          src="http://distillery.s3.amazonaws.com/profiles/test.jpg"
          width={undefined}
        />
        <h3>
          P.J
           
          <small>
            neighborhood999
          </small>
        </h3>
        <h4>
          999
           則貼文
           | 
          999
           位粉絲
           | 
          追蹤 
          999
           人
        </h4>
        <small>
          A guy who write some code.
        </small>
      </div>
      <div
        className="four wide column"
      />
    </div>
  </div>
  <div
    className="ui divider"
  />
  <div
    className="ui stackable grid"
    style={
      Object {
        "marginTop": "50px",
      }
    }
  >
    <div
      className="ui active centered inline loader"
    />
  </div>
</div>
`;
